core/assignment.go

```diff
@@ -149,38 +138,87 @@ func (c *StdAssignmentCoordinator) GetOperatorAssignment(state *OperatorState, q
 	return assignment, info, nil
 }
 
-func (c *StdAssignmentCoordinator) GetMinimumChunkLength(numOperators, blobLength, quantizationFactor uint, quorumThreshold, adversaryThreshold uint8) (uint, error) {
+func (c *StdAssignmentCoordinator) ValidateChunkLength(state *OperatorState, header *BlobHeader, quorum QuorumID) (bool, error) {
 
-	if adversaryThreshold >= quorumThreshold {
-		return 0, errors.New("invalid header: quorum threshold does not exceed adversary threshold")
+	info := header.QuorumInfos[quorum]
+
+	// Check that the chunk length meets the minimum requirement
+	if info.ChunkLength < minChunkLength {
+		return false, ErrChunkLengthTooSmall
 	}
 
-	numSys := roundUpDivide(uint(quorumThreshold-adversaryThreshold)*numOperators*quantizationFactor, PercentMultiplier)
-	chunkLength := roundUpDivide(blobLength, numSys)
-	return chunkLength, nil
+	// Get minimum stake amont
+	minStake := state.Totals[quorum].Stake
+	for _, r := range state.Operators[quorum] {
+		if r.Stake.Cmp(minStake) < 0 {
+			minStake = r.Stake
+		}
+	}
+
+	totalStake := state.Totals[quorum].Stake
+	if info.ChunkLength != minChunkLength {
+
+		num := new(big.Int).Mul(big.NewInt(2*int64(header.Length*percentMultiplier)), minStake)
+		denom := new(big.Int).Mul(big.NewInt(int64(info.QuorumThreshold-info.AdversaryThreshold)), totalStake)
+		maxChunkLength := uint(roundUpDivideBig(num, denom).Uint64())
+
+		maxChunkLength2 := roundUpDivide(2*header.Length*percentMultiplier, maxRequiredNumChunks*uint(info.QuorumThreshold-info.AdversaryThreshold))
+
+		if maxChunkLength < maxChunkLength2 {
+			maxChunkLength = maxChunkLength2
+		}
+
+		maxChunkLength = uint(nextPowerOf2(uint64(maxChunkLength)))
+
+		if info.ChunkLength > maxChunkLength {
+			fmt.Println("maxChunkLength", maxChunkLength, "info.ChunkLength", info.ChunkLength)
+			return false, ErrChunkLengthTooLarge
+		}
+
+	}
+
+	return true, nil
 
 }
 
-func (c *StdAssignmentCoordinator) GetChunkLengthFromHeader(state *OperatorState, header *BlobQuorumInfo) (uint, error) {
+func (c *StdAssignmentCoordinator) CalculateChunkLength(state *OperatorState, blobLength uint, param *SecurityParam) (uint, error) {
+
+	quorum := param.QuorumID
+	numOperators := len(state.Operators[quorum])
+
+	// Get minimum stake amont
+	minStake := state.Totals[quorum].Stake
+	for _, r := range state.Operators[quorum] {
+		if r.Stake.Cmp(minStake) < 0 {
+			minStake = r.Stake
+		}
+	}
+
+	totalStake := state.Totals[quorum].Stake
+
+	numChunks := int(roundUpDivideBig(totalStake, minStake).Uint64())
 
-	// Validate the chunk length
-	numOperators := uint(len(state.Operators[header.QuorumID]))
-	chunkLength := header.EncodedBlobLength / (header.QuantizationFactor * numOperators)
+	if numChunks > maxRequiredNumChunks-numOperators {
+		numChunks = maxRequiredNumChunks - numOperators
+	}
+
```

Comment:
Should numChunks be rounded up to nextPowerOf2 here?

Reply:
Not required actually. The actual number of chunks sent to encoder is calculated separately, and is rounded up to next power. That parameter is independent from this one.